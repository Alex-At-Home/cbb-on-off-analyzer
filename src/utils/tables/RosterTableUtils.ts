// Lodash
import _ from "lodash";

// Util imports
import { RatingUtils } from "../stats/RatingUtils";
import { PositionUtils } from "../stats/PositionUtils";
import { LineupUtils } from "../stats/LineupUtils";
import { LuckUtils, OffLuckAdjustmentDiags, DefLuckAdjustmentDiags, LuckAdjustmentBaseline } from "../stats/LuckUtils";

/** Object marshalling logic for roster tables */
export class RosterTableUtils {

  /** Build a lookup map of the roster by their code, eg AaWiggins for "Wiggins, Aaron" (teamSeasonLookup only needed if includePosCat is true) */
  static buildRosterTableByCode(players: Array<any>, rosterInfo: Record<string, any> | undefined, includePosCat?: boolean, teamSeasonLookup?: string) {
    return _.chain(players).map(p => {
      const code = p.player_array?.hits?.hits?.[0]?._source?.player?.code || p.key;
      const rosterEntry = rosterInfo?.[code];
      // Ugly but we mutate p here:
      if  (rosterEntry) p.roster = rosterEntry;

      //TODO: do I want this one, or the baseline or ???
      if (includePosCat) {
        const [ posConfs, posConfsDiags ] = PositionUtils.buildPositionConfidences(p, rosterEntry?.height_in);
        const [ pos, posDiags ] = PositionUtils.buildPosition(posConfs, posConfsDiags.confsNoHeight, p, teamSeasonLookup || "");
        p.posClass = _.values(posConfs);
        p.role = pos;
      }

      return [ code, p ];
    }).fromPairs().value();

  }
};
