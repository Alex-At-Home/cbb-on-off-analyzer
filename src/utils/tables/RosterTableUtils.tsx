// React
import { ReactNode } from 'react';

// Lodash
import _ from "lodash";

// Util imports
import { PositionUtils } from "../stats/PositionUtils";
import { GenericTableOps } from "../../components/GenericTable";
import { PlayerCode, IndivStatSet } from '../StatModels';

/** Object marshalling logic for roster tables */
export class RosterTableUtils {

  /** Build a lookup map of the roster by their code, eg AaWiggins for "Wiggins, Aaron" (teamSeasonLookup only needed if includePosCat is true) */
  static buildRosterTableByCode(
    players: Array<any>, rosterInfo: Record<string, any> | undefined, includePosCat?: boolean, teamSeasonLookup?: string
  ): Record<PlayerCode, IndivStatSet>  {
    return _.chain(players).map(p => {
      const code = p.player_array?.hits?.hits?.[0]?._source?.player?.code || p.key;
      const rosterEntry = rosterInfo?.[code];
      // Ugly but we mutate p here:
      if  (rosterEntry) p.roster = rosterEntry;

      //TODO: do I want this one, or the baseline or ???
      if (includePosCat) {
        const [ posConfs, posConfsDiags ] = PositionUtils.buildPositionConfidences(p, rosterEntry?.height_in);
        const [ pos, posDiags ] = PositionUtils.buildPosition(posConfs, posConfsDiags.confsNoHeight, p, teamSeasonLookup || "");
        p.posClass = _.values(posConfs);
        p.role = pos;
      }

      return [ code, p ];
    }).fromPairs().value();
  }

  /** Build a table row that sits between the header and the data and gives some colloquial help text */
  static buildInformationalSubheader(calcRapm: boolean, expandedView: boolean) {

    const brace = (color: string, widthPct: number = 100) => <hr style={{
      height: "2px", width: `${widthPct}%`, borderWidth: "0", color: color, backgroundColor: color,
      marginBottom: "0px", paddingBottom: "0px",
      marginTop: "-10px", paddingTop: "0px",
      borderRadius: "5px"
    }}/>;
  
    const braceText = (text: ReactNode, color: string, braceWidth: number = 100) => <div><p style={{color: color}}>
      <i>{text}</i>
    </p>{brace(color, braceWidth)}</div>;
    
    const whereIsAssistedPct = expandedView ? "bottom row is assisted%" : "tooltip shows assisted%";

    return [ 
      GenericTableOps.buildSubHeaderRow([
        ["", 2],
        [braceText("The higher 'Usg' (avg 20) the harder it is to keep ORtg high", "blue"), 2],
        calcRapm ? 
          [braceText(
            <p style={{width: "120%"}}>2x {expandedView?`O/D`:`Off`} Ratings as pts/100 above D1 average (the 2 form a good range)</p>, 
            "purple", 75), 3] 
          : 
          [braceText(
            <p style={{width: "160%"}}>{expandedView?`O/D Ratings`:`Off Rating`} as pts/100 above D1 average</p>, 
            "purple"), 2],
          //TODO: his works weirdly .. colSpan wrong but text right on refresh...
          //(for now I've worked around it by not saving the state of "+ Info")
        ["", 1],
        [braceText("Ball-handling", "brown"), 2],
        ["", expandedView ? 1 : 2], //(OR and DR are both on the top row if not expanded mode)
        [braceText(`Scoring signature (FTR is how often player is fouled shooting, ${whereIsAssistedPct})`, "teal"), 5],
        ["", 1],
        [braceText("Scoring quality", "green"), 4],
        ["", 1],
      ], "small centerAlignCol bottomAlignCol")
    ];
  }
};
