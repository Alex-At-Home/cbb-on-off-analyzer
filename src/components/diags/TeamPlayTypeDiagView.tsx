// React imports:
import React, { useState } from "react";

// Next imports:
import { NextPage } from "next";

import _ from "lodash";

// Bootstrap imports:

import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Tooltip from "react-bootstrap/Tooltip";
import OverlayTrigger from "react-bootstrap/OverlayTrigger";

// Utils
import {
  PosFamilyNames,
  PlayTypeUtils,
  TargetAssistInfo,
} from "../../utils/stats/PlayTypeUtils";
import { PlayTypeDiagUtils } from "../../utils/tables/PlayTypeDiagUtils";

// Component imports
import GenericTable, { GenericTableOps } from "../GenericTable";
import {
  IndivStatSet,
  RosterStatsByCode,
  StatModels,
  TeamStatSet,
} from "../../utils/StatModels";
import TeamPlayTypeDiagRadar from "./TeamPlayTypeDiagRadar";
import { DivisionStatsCache } from "../../utils/tables/GradeTableUtils";

type Props = {
  title: string;
  players: Array<IndivStatSet>;
  rosterStatsByCode: RosterStatsByCode;
  teamStats: TeamStatSet;
  avgEfficiency: number;
  quickSwitchOptions?: Props[];
  showGrades: string;
  grades?: DivisionStatsCache;
  showHelp: boolean;
};
const TeamPlayTypeDiagView: React.FunctionComponent<Props> = ({
  title,
  players: playersIn,
  rosterStatsByCode,
  teamStats: teamStatsIn,
  avgEfficiency,
  quickSwitchOptions,
  showGrades,
  grades,
  showHelp,
}) => {
  const [csvData, setCsvData] = useState<object[]>([]);
  const [quickSwitch, setQuickSwitch] = useState<string | undefined>(undefined);
  const [quickSwitchTimer, setQuickSwitchTimer] = useState<
    NodeJS.Timer | undefined
  >(undefined);
  const players =
    (quickSwitch
      ? _.find(quickSwitchOptions || [], (opt) => opt.title == quickSwitch)
          ?.players
      : playersIn) || [];
  const teamStats =
    (quickSwitch
      ? _.find(quickSwitchOptions || [], (opt) => opt.title == quickSwitch)
          ?.teamStats
      : teamStatsIn) || StatModels.emptyTeam();

  const [tableType, setTableType] = useState<"scoring" | "usage" | "breakdown">(
    "breakdown"
  );

  return React.useMemo(() => {
    const [reorderedPosVsPosAssistNetwork, maybeExtraNetwork] = _.thru(
      tableType,
      () => {
        if (tableType == "scoring") {
          return [
            PlayTypeUtils.buildCategorizedAssistNetworks(
              "scoringPlaysPct",
              false,
              players,
              rosterStatsByCode,
              teamStats
            ),
            undefined,
          ];
        } else if (tableType == "usage") {
          const network1 = PlayTypeUtils.buildCategorizedAssistNetworks(
            "playsPct",
            true,
            players,
            rosterStatsByCode,
            teamStats
          );
          const network2 = PlayTypeUtils.buildCategorizedAssistNetworks(
            "pointsPer100",
            true,
            players,
            rosterStatsByCode,
            teamStats
          );

          return [network1, network2];
        } else {
          return [{}, undefined];
        }
      }
    );

    const tooltipBuilder = (id: string, title: string, tooltip: string) => (
      <OverlayTrigger
        placement="auto"
        overlay={<Tooltip id={id + "Tooltip"}>{tooltip}</Tooltip>}
      >
        <i>{title}</i>
      </OverlayTrigger>
    );

    const rawAssistTableData = (
      tableType == "scoring"
        ? [
            GenericTableOps.buildTextRow(
              <Row>
                <Col xs={3}></Col>
                <Col xs={3} className="d-flex justify-content-center">
                  <i>
                    <span>Scored / Assisted By:</span>
                  </i>
                </Col>
                <Col xs={6} className="d-flex justify-content-center">
                  <i>
                    <span>Assists:</span>
                  </i>
                </Col>
              </Row>
            ),
          ]
        : []
    ).concat(
      _.chain(reorderedPosVsPosAssistNetwork)
        .toPairs()
        .flatMap((kv, ix) => {
          const posTitle = kv[0];
          const assistInfo = kv[1].assists;
          const otherInfo = kv[1].other;

          const extraAssistInfo = maybeExtraNetwork?.[posTitle]?.assists;
          const extraOtherInfo = maybeExtraNetwork?.[posTitle]?.other;

          const asStatSet = (
            targetAssistInfo: TargetAssistInfo | undefined
          ): IndivStatSet =>
            (targetAssistInfo || {}) as unknown as IndivStatSet;

          return [
            GenericTableOps.buildDataRow(
              {
                title: <b>{_.capitalize(posTitle)} from/to:</b>,
              },
              GenericTableOps.defaultFormatter,
              GenericTableOps.defaultCellMeta
            ),
            GenericTableOps.buildDataRow(
              {
                ...PlayTypeDiagUtils.buildInfoRow(
                  PlayTypeUtils.enrichUnassistedStats(
                    otherInfo[0]!,
                    ix,
                    rosterStatsByCode
                  ),
                  asStatSet(extraOtherInfo?.[0])
                ),
                title: tooltipBuilder(
                  "unassisted",
                  "Unassisted",
                  `All scoring plays where the ${posTitle} was unassisted (includes FTs which can never be assisted). ` +
                    (tableType == "usage"
                      ? "(Half-court only)"
                      : "(Includes half court, scramble, and transitions)")
                ),
              },
              GenericTableOps.defaultFormatter,
              GenericTableOps.defaultCellMeta
            ),
            GenericTableOps.buildDataRow(
              {
                ...PlayTypeDiagUtils.buildInfoRow(
                  otherInfo[1]!,
                  asStatSet(extraOtherInfo?.[1])
                ),
                title: tooltipBuilder(
                  "assist",
                  "Assist totals:",
                  `All plays where the  ${posTitle} was assisted (left half) or provided the assist (right half). ` +
                    "The 3 rows below break down assisted plays according to the positional category of the assister/assistee. " +
                    (tableType == "usage"
                      ? "(Half-court only)"
                      : "(Includes half court, scramble, and transitions)")
                ),
              },
              GenericTableOps.defaultFormatter,
              GenericTableOps.defaultCellMeta
            ),
          ]
            .concat(
              GenericTableOps.buildRowSeparator(),
              assistInfo
                .map((info: any, index: number) =>
                  PlayTypeDiagUtils.buildInfoRow(
                    info,
                    asStatSet(extraAssistInfo?.[index])
                  )
                )
                .map((info: any) =>
                  GenericTableOps.buildDataRow(
                    {
                      ...info,
                      title: (
                        <span>
                          <i>{_.capitalize(PosFamilyNames[info.order])}</i>
                        </span>
                      ),
                    },
                    GenericTableOps.defaultFormatter,
                    GenericTableOps.defaultCellMeta
                  )
                )
            )
            .concat([
              GenericTableOps.buildRowSeparator(),
              GenericTableOps.buildDataRow(
                {
                  ...PlayTypeDiagUtils.buildInfoRow(
                    otherInfo[2]!,
                    asStatSet(extraOtherInfo?.[2])
                  ),
                  title: tooltipBuilder(
                    "trans",
                    "In transition",
                    "All plays (assisted or unassisted) that are classified as 'in transition', normally shots taken rapidly after a rebound, miss, or make in the other direction."
                  ),
                },
                GenericTableOps.defaultFormatter,
                GenericTableOps.defaultCellMeta
              ),
              GenericTableOps.buildDataRow(
                {
                  ...PlayTypeDiagUtils.buildInfoRow(
                    otherInfo[3]!,
                    asStatSet(extraOtherInfo?.[3])
                  ),
                  title: tooltipBuilder(
                    "scramble",
                    "Scrambles after RB",
                    "All plays (assisted or unassisted) that occur in the aftermath of an offensive rebound, where the offense does not get reset before scoring. " +
                      "Examples are putbacks (unassisted) or tips to other players (assisted)"
                  ),
                },
                GenericTableOps.defaultFormatter,
                GenericTableOps.defaultCellMeta
              ),
              GenericTableOps.buildRowSeparator(),
            ]);
        })
        .value()
    );

    const maybeBold = (boldIf: String, el: React.ReactElement) => {
      if (boldIf == tableType) {
        return <b>{el}</b>;
      } else {
        return el;
      }
    };

    const scoringToggle = maybeBold(
      "scoring",
      <a
        href="#"
        onClick={(event) => {
          event.preventDefault();
          setTableType("scoring");
        }}
      >
        Scoring
      </a>
    );
    const usageToggle = maybeBold(
      "usage",
      <a
        href="#"
        onClick={(event) => {
          event.preventDefault();
          setTableType("usage");
        }}
      >
        Usage
      </a>
    );

    const breakdownToggle = maybeBold(
      "breakdown",
      <a
        href="#"
        onClick={(event) => {
          event.preventDefault();
          setTableType("breakdown");
        }}
      >
        Play Types
      </a>
    );

    return (
      <span>
        {/*JSON.stringify(_.chain(teamStats).toPairs().filter(kv => kv[0].indexOf("trans") >= 0).values(), tidyNumbers, 3)*/}
        <br />
        {title
          ? PlayTypeDiagUtils.buildQuickSwitchOptions(
              title,
              quickSwitch,
              quickSwitchOptions,
              setQuickSwitch,
              quickSwitchTimer,
              setQuickSwitchTimer
            )
          : undefined}
        <span>
          ({scoringToggle} // {usageToggle} // {breakdownToggle}){" "}
          {tableType == "breakdown"
            ? PlayTypeDiagUtils.buildLegend("[LEGEND]")
            : null}
          {tableType == "breakdown" && grades ? <span> | </span> : null}
          {tableType == "breakdown" && grades
            ? PlayTypeDiagUtils.buildCsvDownload(
                "[CSV]",
                `play_types_${title}`,
                csvData,
                () => {
                  const playStyleData: object[] =
                    PlayTypeDiagUtils.buildTeamStyleBreakdownData(
                      title,
                      true,
                      "",
                      {
                        on: [],
                        off: [],
                        other: [],
                        baseline: players,
                        global: [],
                      },
                      {
                        on: StatModels.emptyTeam(),
                        off: StatModels.emptyTeam(),
                        other: [],
                        baseline: teamStats,
                        global: StatModels.emptyTeam(),
                      },
                      avgEfficiency,
                      grades,
                      false,
                      undefined,
                      rosterStatsByCode
                    );
                  setCsvData(playStyleData);
                }
              )
            : null}
        </span>
        <br />
        {tableType == "breakdown" ? (
          <TeamPlayTypeDiagRadar
            players={players}
            rosterStatsByCode={rosterStatsByCode}
            teamStats={teamStats}
            avgEfficiency={avgEfficiency}
            showGrades={showGrades}
            grades={grades}
            showHelp={showHelp}
            quickSwitchOverride={quickSwitch}
          />
        ) : (
          <Container>
            <Col xs={10}>
              <GenericTable
                responsive={false}
                tableCopyId="teamAssistNetworks"
                tableFields={PlayTypeDiagUtils.rawAssistTableFields(
                  false,
                  true,
                  tableType,
                  tableType == "usage"
                )}
                tableData={rawAssistTableData}
              />
            </Col>
          </Container>
        )}
      </span>
    );
  }, [
    players,
    tableType,
    grades,
    showGrades,
    teamStats,
    quickSwitch,
    quickSwitchTimer,
    csvData,
  ]);
};
export default TeamPlayTypeDiagView;
