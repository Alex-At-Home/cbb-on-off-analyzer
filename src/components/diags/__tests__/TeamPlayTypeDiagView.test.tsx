import renderer from "react-test-renderer";
import React from "react";
import _ from "lodash";
import TeamPlayTypeDiagView from "../TeamPlayTypeDiagView";
import { samplePlayerStatsResponse } from "../../../sample-data/samplePlayerStatsResponse";
import { sampleTeamStatsResponse } from "../../../sample-data/sampleTeamStatsResponse";
import { GameFilterParams } from "../../../utils/FilterModels";
import { RosterTableUtils } from "../../../utils/tables/RosterTableUtils";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { IndivStatSet, TeamStatSet } from "../../../utils/StatModels";

describe("TeamPlayTypeDiagView", () => {
  const testData = {
    on:
      samplePlayerStatsResponse.responses[0].aggregations?.tri_filter?.buckets
        ?.on?.player?.buckets || [],
    off:
      samplePlayerStatsResponse.responses[0].aggregations?.tri_filter?.buckets
        ?.off?.player?.buckets || [],
    baseline:
      samplePlayerStatsResponse.responses[0].aggregations?.tri_filter?.buckets
        ?.baseline?.player?.buckets || [],
    error_code: undefined,
  };
  const teamData = _.assign(
    sampleTeamStatsResponse.responses[0].aggregations.tri_filter.buckets as {
      on: any;
      off: any;
      baseline: any;
    },
    { global: {}, onOffMode: true }
  ) as unknown as TeamStatSet;
  const teamSeasonLookup = "Men_Maryland_2018/9";
  const rosterStatsByCode = RosterTableUtils.buildRosterTableByCode(
    testData.on,
    undefined,
    true,
    teamSeasonLookup
  );
  const players = testData.on.map((p) => {
    //inject pos class into data
    const code = p.player_array?.hits?.hits?.[0]?._source?.player?.code;
    return {
      ...p,
      posClass: code ? rosterStatsByCode[code]?.posClass || "WG" : "WG",
    };
  }) as unknown as IndivStatSet[];
  test("TeamPlayTypeDiagView - should create snapshot (!details, help)", () => {
    const wrapper = shallow(
      <TeamPlayTypeDiagView
        title={"Title"}
        players={players}
        rosterStatsByCode={rosterStatsByCode}
        teamStats={teamData}
        showGrades="false"
        showHelp={false}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  test("TeamPlayTypeDiagView - should create snapshot (!help)", () => {
    const wrapper = shallow(
      <TeamPlayTypeDiagView
        title={"Title"}
        players={players}
        rosterStatsByCode={rosterStatsByCode}
        teamStats={teamData}
        showGrades="false"
        showHelp={false}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
