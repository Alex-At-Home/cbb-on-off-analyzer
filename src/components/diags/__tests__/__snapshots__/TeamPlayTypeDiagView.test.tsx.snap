// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamPlayTypeDiagView TeamPlayTypeDiagView - should create snapshot (!details, help) 1`] = `
<span>
  <div>
    <span
      style={
        Object {
          "display": "inline-block",
          "whiteSpace": "nowrap",
        }
      }
    >
      <b>
        Analysis: [
        Title
        ]
      </b>
    </span>
  </div>
  <Container
    className="mt-2"
    fluid={false}
  >
    <Col
      xs={10}
    >
      <GenericTable
        responsive={false}
        tableCopyId="teamAssistNetworks"
        tableData={Array []}
        tableFields={
          Object {
            "sepsource": GenericTableColProps {
              "className": "",
              "colName": "",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "",
              "widthUnits": 0.75,
            },
            "sepsourceast": GenericTableColProps {
              "className": "",
              "colName": "",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "",
              "widthUnits": 0.25,
            },
            "septarget": GenericTableColProps {
              "className": "",
              "colName": "",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "",
              "widthUnits": 0.75,
            },
            "septarget3p": GenericTableColProps {
              "className": "",
              "colName": "",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "",
              "widthUnits": 0.125,
            },
            "septargetmid": GenericTableColProps {
              "className": "",
              "colName": "",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "",
              "widthUnits": 0.125,
            },
            "septargetrim": GenericTableColProps {
              "className": "",
              "colName": "",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "",
              "widthUnits": 0.125,
            },
            "source_3p_ast": GenericTableColProps {
              "className": "",
              "colName": "3P ",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "% of team total possessions of this shot type assisted BY the specified row (positional role)",
              "widthUnits": 2,
            },
            "source_mid_ast": GenericTableColProps {
              "className": "",
              "colName": "Mid ",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "% of team total possessions of this shot type assisted BY the specified row (positional role)",
              "widthUnits": 2,
            },
            "source_rim_ast": GenericTableColProps {
              "className": "",
              "colName": "Rim ",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "% of team total possessions of this shot type assisted BY the specified row (positional role)",
              "widthUnits": 2,
            },
            "source_sf": GenericTableColProps {
              "className": "",
              "colName": "SF",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "% of team total possessions ending in a trip to the FT line",
              "widthUnits": 2,
            },
            "target_3p_ast": GenericTableColProps {
              "className": "",
              "colName": "3P AST% ",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "% of total assists to the specified row (positional role) for this shot type",
              "widthUnits": 2,
            },
            "target_3p_efg": GenericTableColProps {
              "className": "",
              "colName": "eFG",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "The season eFG% of this shot type / row (positional role)",
              "widthUnits": 2,
            },
            "target_ast": GenericTableColProps {
              "className": "",
              "colName": "AST",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "% of team total possessions ending with an assist FOR the specified row (positional role)",
              "widthUnits": 2,
            },
            "target_mid_ast": GenericTableColProps {
              "className": "",
              "colName": "Mid AST% ",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "% of total assists to the specified row (positional role) for this shot type",
              "widthUnits": 2,
            },
            "target_mid_efg": GenericTableColProps {
              "className": "",
              "colName": "eFG",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "The season eFG% of this shot type / row (positional role)",
              "widthUnits": 2,
            },
            "target_rim_ast": GenericTableColProps {
              "className": "",
              "colName": "Rim AST% ",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "% of total assists to the specified row (positional role) for this shot type",
              "widthUnits": 2,
            },
            "target_rim_efg": GenericTableColProps {
              "className": "",
              "colName": "eFG",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "The season eFG% of this shot type / row (positional role)",
              "widthUnits": 2,
            },
            "title": GenericTableColProps {
              "className": "",
              "colName": "",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": true,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "",
              "widthUnits": 8,
            },
          }
        }
      />
    </Col>
  </Container>
</span>
`;

exports[`TeamPlayTypeDiagView TeamPlayTypeDiagView - should create snapshot (!help) 1`] = `
<span>
  <div>
    <span
      style={
        Object {
          "display": "inline-block",
          "whiteSpace": "nowrap",
        }
      }
    >
      <b>
        Analysis: [
        Title
        ]
      </b>
    </span>
  </div>
  <Container
    className="mt-2"
    fluid={false}
  >
    <Col
      xs={10}
    >
      <GenericTable
        responsive={false}
        tableCopyId="teamAssistNetworks"
        tableData={Array []}
        tableFields={
          Object {
            "sepsource": GenericTableColProps {
              "className": "",
              "colName": "",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "",
              "widthUnits": 0.75,
            },
            "sepsourceast": GenericTableColProps {
              "className": "",
              "colName": "",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "",
              "widthUnits": 0.25,
            },
            "septarget": GenericTableColProps {
              "className": "",
              "colName": "",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "",
              "widthUnits": 0.75,
            },
            "septarget3p": GenericTableColProps {
              "className": "",
              "colName": "",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "",
              "widthUnits": 0.125,
            },
            "septargetmid": GenericTableColProps {
              "className": "",
              "colName": "",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "",
              "widthUnits": 0.125,
            },
            "septargetrim": GenericTableColProps {
              "className": "",
              "colName": "",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "",
              "widthUnits": 0.125,
            },
            "source_3p_ast": GenericTableColProps {
              "className": "",
              "colName": "3P ",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "% of team total possessions of this shot type assisted BY the specified row (positional role)",
              "widthUnits": 2,
            },
            "source_mid_ast": GenericTableColProps {
              "className": "",
              "colName": "Mid ",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "% of team total possessions of this shot type assisted BY the specified row (positional role)",
              "widthUnits": 2,
            },
            "source_rim_ast": GenericTableColProps {
              "className": "",
              "colName": "Rim ",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "% of team total possessions of this shot type assisted BY the specified row (positional role)",
              "widthUnits": 2,
            },
            "source_sf": GenericTableColProps {
              "className": "",
              "colName": "SF",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "% of team total possessions ending in a trip to the FT line",
              "widthUnits": 2,
            },
            "target_3p_ast": GenericTableColProps {
              "className": "",
              "colName": "3P AST% ",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "% of total assists to the specified row (positional role) for this shot type",
              "widthUnits": 2,
            },
            "target_3p_efg": GenericTableColProps {
              "className": "",
              "colName": "eFG",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "The season eFG% of this shot type / row (positional role)",
              "widthUnits": 2,
            },
            "target_ast": GenericTableColProps {
              "className": "",
              "colName": "AST",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "% of team total possessions ending with an assist FOR the specified row (positional role)",
              "widthUnits": 2,
            },
            "target_mid_ast": GenericTableColProps {
              "className": "",
              "colName": "Mid AST% ",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "% of total assists to the specified row (positional role) for this shot type",
              "widthUnits": 2,
            },
            "target_mid_efg": GenericTableColProps {
              "className": "",
              "colName": "eFG",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "The season eFG% of this shot type / row (positional role)",
              "widthUnits": 2,
            },
            "target_rim_ast": GenericTableColProps {
              "className": "",
              "colName": "Rim AST% ",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "% of total assists to the specified row (positional role) for this shot type",
              "widthUnits": 2,
            },
            "target_rim_efg": GenericTableColProps {
              "className": "",
              "colName": "eFG",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": false,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "The season eFG% of this shot type / row (positional role)",
              "widthUnits": 2,
            },
            "title": GenericTableColProps {
              "className": "",
              "colName": "",
              "colorPicker": [Function],
              "formatter": [Function],
              "isTitle": true,
              "missingData": undefined,
              "rowSpan": [Function],
              "toolTip": "",
              "widthUnits": 8,
            },
          }
        }
      />
    </Col>
  </Container>
</span>
`;
